Add Pancakeswap NFT Events

thb030100 opened the pull request:
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test

dune-eng commented:
Workflow run id 3640654290 approved.

dune-eng commented:
Workflow run id 3640654291 approved.

dune-eng commented:
Workflow run id 3644726959 approved.

dune-eng commented:
Workflow run id 3645025551 approved.

0xRobin commented on models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql:
```diff
@@ -0,0 +1,89 @@
+        {% if is_incremental() %}
+        AND prices.minute >= date_trunc("day", now() - interval '1 week')
+        {% endif %}
+    LEFT JOIN {{ source('bnb','transactions') }} bt ON bt.hash=events.tx_hash
```
This can be made an inner join for better performance
```suggestion
    INNER JOIN {{ source('bnb','transactions') }} bt ON bt.hash=events.tx_hash
```

0xRobin commented on models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql:
```diff
@@ -0,0 +1,89 @@
+        buyer AS buyer,
+        seller AS seller,
+        askPrice AS amount_raw,
+        '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c' currency_contract,
+        'BNB' currency_symbol,
```
```suggestion
        '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c' AS currency_contract,
        'BNB' AS currency_symbol,
```

0xRobin commented on models/pancakeswap/bnb/pancakeswap_bnb_schema.yml:
```diff
@@ -94,4 +94,89 @@ models:
-        description: "Index of the corresponding trade event"
\ No newline at end of file
+        description: "Index of the corresponding trade event"
+  - name: pancakeswap_bnb_nft_market_v1_event
+    meta:
+      blockchain: bnb
+      project: element
+      contributors: thb0301
+    config:
+      tags: ['bnb','pancakeswap','events']
+    description: >
+        Pancakeswap events on BNB
```
We should also add tests here:
```yml
  tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - block_date
            - unique_trade_id
      - check_seed:
          seed_file: ref('pancakeswap_nft_trades_samples')
          match_columns:
            - block_time
            - tx_hash
          check_columns:
            - token_id
            - amount_original
            - buyer
            - seller
```

0xRobin commented on models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql:
```diff
@@ -0,0 +1,89 @@
+        CAST(0 AS DOUBLE) AS royalty_fee_percentage,
+        0 AS royalty_fee_receive_address,
+        0 AS royalty_fee_currency_symbol,
+        events.blockchain || events.project || events.version || events.tx_hash || events.seller  || events.buyer || events.nft_contract_address || events.token_id AS unique_trade_id
```
we're trying to minimize the amount of columns used here.
`block_number - tx_hash - evt_index` will do in this case.
```suggestion
        events.block_number || events.tx_hash || events.evt_index AS unique_trade_id
```
You'll have to add the evt_index in the events CTE above.

0xRobin reviewed (changes requested):
Thanks for the PR, I've added some minor suggestions.
Main thing is we'd like to see a seed file with a couple example trades in them to setup a test.
You can add that under `seeds/pancakeswap/bnb/pancakeswap_nft_trades_samples.csv`.
Then you'll also have to update the `dbt_project.yml` file to include the seed.

Let me know if you have questions.

dune-eng commented:
Workflow run id 3649973300 approved.

dune-eng commented:
Workflow run id 3649973301 approved.

dune-eng commented:
Workflow run id 3649994929 approved.

dune-eng commented:
Workflow run id 3649994928 approved.

dune-eng commented:
Workflow run id 3649996711 approved.

dune-eng commented:
Workflow run id 3649996707 approved.

dune-eng commented:
Workflow run id 3650017945 approved.

dune-eng commented:
Workflow run id 3650433225 approved.

dune-eng commented:
Workflow run id 3650433215 approved.

thb030100 commented:
> Thanks for the PR, I've added some minor suggestions. Main thing is we'd like to see a seed file with a couple example trades in them to setup a test. You can add that under `seeds/pancakeswap/bnb/pancakeswap_nft_trades_samples.csv`. Then you'll also have to update the `dbt_project.yml` file to include the seed.
> 
> Let me know if you have questions.

I have added the seed file, can you check! Thanks

dune-eng commented:
Workflow run id 3654025232 approved.

dune-eng commented:
Workflow run id 3654025233 approved.

dune-eng commented:
Workflow run id 3654025235 approved.

dune-eng commented:
Workflow run id 3654026978 approved.

dune-eng commented:
Workflow run id 3654026975 approved.

dune-eng commented:
Workflow run id 3654027953 approved.

dune-eng commented:
Workflow run id 3654027954 approved.

dune-eng commented:
Workflow run id 3654035619 approved.

dune-eng commented:
Workflow run id 3654035621 approved.

dune-eng commented:
Workflow run id 3654044260 approved.

dune-eng commented:
Workflow run id 3654044263 approved.

dune-eng commented:
Workflow run id 3654048931 approved.

dune-eng commented:
Workflow run id 3654050158 approved.

dune-eng commented:
Workflow run id 3654056049 approved.

dune-eng commented:
Workflow run id 3654056058 approved.

dune-eng commented:
Workflow run id 3654079310 approved.

dune-eng commented:
Workflow run id 3654079312 approved.

dune-eng commented:
Workflow run id 3654084677 approved.

dune-eng commented:
Workflow run id 3672759632 approved.

dune-eng commented:
Workflow run id 3672759631 approved.

dune-eng commented:
Workflow run id 3672763480 approved.

dune-eng commented:
Workflow run id 3672763477 approved.

0xRobin commented:
Looks like the test samples are not passing, @thb030100 can you check?

dune-eng commented:
Workflow run id 3702819551 approved.

dune-eng commented:
Workflow run id 3702819553 approved.

thb030100 commented:
> 

I don't know why the test fail. Can you check for me?

0xRobin commented:
Yes I'll have a look

thb030100 commented:
> Yes I'll have a look

Have you founded it?

dune-eng commented:
Workflow run id 3729664390 approved.

dune-eng commented:
Workflow run id 3729664391 approved.

dune-eng commented:
Workflow run id 3729737792 approved.

dune-eng commented:
Workflow run id 3729737787 approved.

thb030100 commented:
I temporary comment on the test section since I don't understand clearly how the test works
`with matched_records as (
        select
            seed.block_time as seed_block_time,
            model_sample.block_time as model_block_time,
            
            seed.tx_hash as seed_tx_hash,
            model_sample.tx_hash as model_tx_hash,
            
            seed.token_id as seed_token_id,
            model_sample.token_id as model_token_id ,
            
            seed.amount_original as seed_amount_original,
            model_sample.amount_original as model_amount_original ,
            
            seed.buyer as seed_buyer,
            model_sample.buyer as model_buyer ,
            
            seed.seller as seed_seller,
            model_sample.seller as model_seller 
            from test_data.pancakeswap_nft_trades_samples seed
        left join (
            select
                model.block_time,
                
                model.tx_hash,
                
                model.token_id ,
                
                model.amount_original ,
                
                model.buyer ,
                
                model.seller 
                from  test_data.pancakeswap_nft_trades_samples seed
            inner join pancakeswap_bnb.events model
                ON 1=1
                    AND seed.block_time = model.block_time
                    
                    AND seed.tx_hash = model.tx_hash
                    ) model_sample
        ON 1=1
            AND seed.block_time = model_sample.block_time
            
            AND seed.tx_hash = model_sample.tx_hash
            WHERE 1=1),
`
Following this code, I guest it compare with pancakeswap_bnb.events which not exists yet. This might be the issue

dune-eng commented:
Workflow run id 3729892221 approved.

dune-eng commented:
Workflow run id 3729892219 approved.

0xRobin commented:
@thb030100 hey I just pushed to re-enable the tests 
(also updated the columns of the model in the schema file)

I'll walk you through how we can evaluate what is happening. 
So the dbt slim ci process will produce temporary tables for both the model and the seed file.
These you can query on Dune. :astonished: 

So if a test is not passing or working as expected:
1. [local] dbt compile and track down the compiled test file
2. check the logs from the dbt slim ci to get the names of the temporary model and seed tables
3. replace the sources in the compiled test with the temporary tables.
4. Run on Dune and investigate the results.

I'll wait for the dbt slim ci to run and give more instructions, we can walk through this together.

0xRobin commented:
CC: @cryptofreedman we should probably describe this process in https://dune.com/docs/spellbook/getting-started/tests/ to help others.

thb030100 commented:
@0xRobin thank you so much for the instruction! I have changed the test file and let see how it goes. 
This pr requests was so tough lol

dune-eng commented:
Workflow run id 3734086889 approved.

dune-eng commented:
Workflow run id 3734086892 approved.

0xRobin commented:
I believe it's not picking the dates in the seed file correctly. The format should be 'YYYY-MM-DD hh:mm:ss'

thb030100 commented:
> I believe it's not picking the dates in the seed file correctly. The format should be 'YYYY-MM-DD hh:mm:ss'

Hey I have changed the format, it's YYYY-MM-DD hh:mm:ss now

2021-09-30 10:06:35

dune-eng commented:
Workflow run id 3740431887 approved.

dune-eng commented:
Workflow run id 3740431899 approved.

0xRobin commented:
@thb030100 looks like tests are passing! :clap: 
I will add some changes for consistency and integrate this model with the general nft.trades one.
After that we are good to merge this!
Thanks for the contribution, the first one is always the hardest. :smile:

thb030100 commented:
> @thb030100 looks like tests are passing! 👏
> I will add some changes for consistency and integrate this model with the general nft.trades one.
> After that we are good to merge this!
> Thanks for the contribution, the first one is always the hardest. 😄

Thank you! I am also adding a few more marketplaces. Will follow this PR as an instruction!

dune-eng commented:
Workflow run id 3746198852 approved.

dune-eng commented:
Workflow run id 3746198851 approved.

dot2dotseurat commented:
Sorry, switching this to do not merge for the time being. I'm checking if this will work on DuneSQL and it will cause nft.trades to be unqueryable. I need to look more closely to see why. 

@thb030100 thank you for making this contribution, we'll get it merged and welcome more contributions. Excellent first contribution by the way!

belen-pruvost commented:
Hey @dot2dotseurat! Yes, the underlying issue here is that `test_schema.git_653e5544_pancakeswap_bnb_nft_events` has a DBFS location instead of an S3 one (see [here](https://dbc-7ee55e77-eabe.cloud.databricks.com/sql/explore/data/hive_metastore/test_schema/git_653e5544_pancakeswap_bnb_nft_events?o=4190327660257548)). Ideally the CI that creates the test schema would need to use something like what we discussed [here](https://dune.height.app/T-22083).

dot2dotseurat commented:
@belen-pruvost ah, makes sense. thanks for checking. 

We can probably go ahead and merge this then.

soispoke reviewed (approved):
lgtm
